### ISSUE TITLE
Each file contains the images for K shots for a specific category. There may be more instances in the images than K; in these cases, we randomly sample K instances.

### ISSUE THREAD

**Issue by Chen-Song** [MegviiDetection/FSCE#13]
Hi, please tell me what does this sentence mean? I am a little confused, thanks

**Comment by Chen-Song** [comment]
Excuse me, where did you randomly sample k instances in your code?

**Comment by Chauncy-Cai** [comment]
(1)For a K-shot task, we can only get K instances of novel classes for training or finetune. If there are two 'chair' in a image, but we only need 1 shot. We will randomly choose one of them instead of both of them.
(2)Sampling k instances randomly should be done before training, you can run "prepare_[coco,voc,lvis]_few_shot.py" file in "FSCE/datasets/" to generate random k instances. After that, you should change your yaml to figure out which training dataset will be used.

**Comment by Chen-Song** [comment]
Thanks for your reply. I run prepare_voc_few_shot.py and find there are more instances than the pre-defined k. For example, in "http://dl.yf.io/fs-det/datasets/vocsplit/seed1/box_10shot_sheep_train.txt", there are seven images while these images contain 24 'sheep' instances.

**Comment by Chen-Song** [comment]
In 'http://dl.yf.io/fs-det/datasets/vocsplit/seed1/box_5shot_tvmonitor_train.txt', there are four images while these images contain six 'tvmonitor' instances. Maybe there is some important information that I have overlooked or I have a wrong understanding of these generated files. Look forward to your reply.

**Comment by Retiina** [comment]
https://github.com/ucbdrive/few-shot-object-detection/issues/89

**Comment by Chen-Song** [comment]
Thanks, “Is this expected behavior or a bug?”

**Comment by Chauncy-Cai** [comment]
I think it's expected behavior. When you run train_net, you should see a chart, which will tell you the number of instances from every class.

**Comment by john2020-210** [comment]
where can I find the configs file in the model_zoo ,as it is a soft link in the source file

**Comment by Chauncy-Cai** [comment]
@john2020-210 It's the output of the build operation. Please delete it, or the build operation may not succeed.

**Comment by john2020-210** [comment]
@Chauncy-Cai Thank you for your reply

**Comment by bsun0802** [comment]
> Thanks for your reply. I run prepare_voc_few_shot.py and find there are more instances than the pre-defined k. For example, in "http://dl.yf.io/fs-det/datasets/vocsplit/seed1/box_10shot_sheep_train.txt", there are seven images while these images contain 24 'sheep' instances.

@Chen-Song The K-shot preparation comes from fsdet. We inspected the instances and have same concern with you. I also noticed fsdet dataloader load 1 instance per image, and ignoring all other annotations. Maybe that's their definition of "few-shot object detection". And actually, we are working on improving the sampling for K-shot, and change the data loader in our v2. Too much details to discuss here, we encourage you to do your own research, and formulate your own K-shot sampling schema.

**Comment by Retiina** [comment]
@bsun0802 I think it is a bug. If we fix seed before random choice, see here:

https://github.com/MegviiDetection/FSCE/blob/94ba3f4737d5e40af795db49b8c6914526c912b6/fsdet/data/datasets/meta_pascal_voc.py#L129-L130

In DDP training different processes may select different instances though the number of instances of a single process is exact K. This is unfair when comparing other method which use exact K shot for training.
If we fix the seed, say `np.random.seed(seed)` before `random_choice`, then each process will select exactly the same instances and the number of overall instances is exact K. The performance will drop a lot.

**Comment by bsun0802** [comment]
This code is from fsdet unchanged. 

I have a different understanding than yours. And I don't think the problem you mention exists, based on my experience. Here is mine:

In my understanding and experience with detectron2, `meta_pascal_voc.py` is used for registering the dataset. Which is executed on 1 cpu and executed only once for any training job,  before any model training (DDP) code. This step register which **exact K-shot** to be used as the training data. And the actual data loader is in the function `build_detection_train_loader`.

**Comment by bsun0802** [comment]
This code is from fsdet unchanged. 

I have a different understanding than yours @Retiina . And I don't think the problem you mention exists, based on my experience. Here is mine:

In my understanding and experience with detectron2, `meta_pascal_voc.py` is used for registering the dataset. Which is executed on 1 cpu and executed only once for any training job,  before any model training (DDP) code. This step register which **exact K-shot** to be used as the training data. And the actual data loader is in the function `build_detection_train_loader`.

**Comment by Retiina** [comment]
@bsun0802 Thanks for your reply. I think you are right.
